### ISSUE TITLE
segfault in nk_group_scrolled_offset_begin

### ISSUE THREAD

**Issue by wheybags** [vurtun/nuklear#477]
The following code will segafult on attempting to access y_offset in nk_group_scrolled_offset_begin.
The issue only seems to be present when compiled as a 64-bit binary.
The code is a minimal example derived from a file browser, hence the weird strings. To reproduce, click the AAA button three times, then scroll down to the point that the "asset_browser_.dnx" button should be visible, at which point it will segfault.
The AAA button simulates changing directories, which are drawn as groups inside a group inside a window.

```c  
static int strsIndex = 0;


nk_flags windowFlags = NK_WINDOW_BORDER | NK_WINDOW_MOVABLE | NK_WINDOW_SCALABLE |
    NK_WINDOW_CLOSABLE | NK_WINDOW_MINIMIZABLE | NK_WINDOW_TITLE;

if(nk_begin(ctx, "bleep bloop", nk_rect(50, 50, 300, 300), windowFlags))
{
    nk_layout_row_dynamic(ctx, 25, 1);


    if (nk_button_label(ctx, "AAA"))
    {
        strsIndex++;
    }

    char* str1[] = { "asset_browser_a" };

    char* str2[] = {
        "asset_browser_.android",
        "asset_browser_.AndroidStudio2.1",
        "asset_browser_.ant",
        "asset_browser_.dnx",
        "asset_browser_.gimp-2.8",
        "asset_browser_.hunter",
        "asset_browser_.nuget",
        "asset_browser_.oracle_jre_usage",
        "asset_browser_.ssh",
        "asset_browser_.thumbnails",
        "asset_browser_.vim",
        "asset_browser_.VirtualBox",
        "asset_browser_.vs",
        "asset_browser_.x2go",
        "asset_browser_Contacts",
        "asset_browser_Desktop",
        "asset_browser_Documents",
        "asset_browser_dotfiles",
        "asset_browser_Downloads",
        "asset_browser_Dropbox",
        "asset_browser_Favorites",
        "asset_browser_Links",
        "asset_browser_Music",
        "asset_browser_Pictures",
        "asset_browser_Saved Games",
        "asset_browser_Searches",
        "asset_browser_src_c",
        "asset_browser_ssh",
        "asset_browser_testtar",
        "asset_browser_Videos",
        "asset_browser_VirtualBox VMs"
    };

    char* str3[] = {
        "asset_browser_Public",
        "asset_browser_wheybags",
    };

    char* str4[] = {
        "asset_browser_.android",
        "asset_browser_.AndroidStudio2.1",
        "asset_browser_.ant",
        "asset_browser_.dnx"
    };

    char** strs[4] = { str1, str2, str3, str4 };
    int lens[4] = {
        sizeof(str1) / sizeof(char*),
        sizeof(str2) / sizeof(char*),
        sizeof(str3) / sizeof(char*),
        sizeof(str4) / sizeof(char*),
    };

    nk_layout_row_dynamic(ctx, 300, 1);

    if(nk_group_begin(ctx, "asset_browser", 0))
    {
        nk_layout_row_static(ctx, 190, 190, 1);

        for(int i = 0; i < lens[strsIndex]; i++)
        {
            char* a = strs[strsIndex][i];

            if (nk_group_begin(ctx, a, NK_WINDOW_NO_SCROLLBAR))
            {
                nk_layout_row_dynamic(ctx, 25, 1);
                nk_button_label(ctx, a);

                nk_group_end(ctx);
            }
        }

        nk_group_end(ctx);
    }
}
nk_end(ctx);
```

The real problem seems to be somewhere in the nk_table implementation, the pointer is null because nk_find_value failed to find it when called from nk_group_begin. In specific, I think it's something to do with corruption over page boundaries, as if I set NK_VALUE_PAGE_CAPACITY to a large value (large enough that we only ever use one page), the problem disappears.
I think this is also why the problem doesn't manifest on 32-bit systems, as NK_VALUE_PAGE_CAPACITY is calculated from the size of various structs (why is this, actually?), and is therefore different.

Also, a side note, it seems that nk_table is using only hashes for storage and lookup, isn't this vulnerable to hash collision?

**Comment by vurtun** [comment]
Hey thanks for the issue. 
> I think it's something to do with corruption over page boundaries, as if I set NK_VALUE_PAGE_CAPACITY to a large value (large enough that we only ever use one page), the problem disappears I think this is also why the problem doesn't manifest on 32-bit systems, as NK_VALUE_PAGE_CAPACITY is calculated from the size of various structs (why is this, actually?), and is therefore different.

Close but not directly.  In the past I stored the size of the table inside the window and never pulled it
into the actual table. So if you actually allocate more than the table the size is from the new table and
therefore does not run over the complete array of values and returns NULL.
Bug should now be fixed in commit: b52a7398949c753764167278a8a6fec32dbac392.

> NK_VALUE_PAGE_CAPACITY is calculated from the size of various structs (why is this, actually?),

All runtime allocated persistent data structures with `nk_panel`, `nk_window` and `nk_table` are put into one union and allocated by a fixed block allocator. However since tables contain two arrays on for keys
and one for values it has to fit roughly into the same size as `nk_panel` and `nk_window`. Thats why
this complex array size macro exists. 

> Also, a side note, it seems that nk_table is using only hashes for storage and lookup, isn't this vulnerable to hash collision?

While I somewhat trust murmur yes that is a problem and a reason why `nk_window`s store their name inside an array and check if correct to make sure no hash collisions happen. However tables are a lot more difficult to ensure to have no collisions. Each entry would have to store a string representation which takes up a lot of memory.  One thing that could be done is having another hash function generate another hash and check if both are correct. But I don't know if it is actually worth it.

**Comment by dumblob** [comment]
>While I somewhat trust murmur yes that is a problem and a reason why `nk_windows` store their name inside an array and check if correct to make sure no hash collisions happen.

If I understand it correctly, you're simply using an array as a "linked list" of "overflows" for each hash collision. This overflow linked list is a standard solution for very good hashing functions (which murmur is) as it's easy to implement and for absolutely majority of cases it's even performant (as lists are usually arrays with very few items - two to five - which are checked in a for loop which in turn is compiled with vectorization and many [other techniques](https://en.wikipedia.org/wiki/Loop_optimization ) which is then nearly as fast as computing a hash for such tiny number of inputs.

> One thing that could be done is having another hash function generate another hash and check if both are correct.

This wouldn't work, because you would need the second hash function to cover exactly all the collision places of the first hash function, which would mathematically mean you actually constructed a general collision-free hash function, which is a nonsense for arbitrary inputs (i.e. inputs about which we don't know absolutely anything).
